File: mobile/lib/utils/transfer_metrics.dart
import 'dart:math';

class TransferMetricsState {
  final int totalBytes;
  final int resumedBytes;
  int transferredBytes;
  final DateTime startedAt;
  DateTime lastSampleAt;
  int lastSampleBytes;
  double? smoothedBytesPerSecond;

  TransferMetricsState({
    required this.totalBytes,
    this.resumedBytes = 0,
    required this.startedAt,
    required this.lastSampleAt,
    required this.lastSampleBytes,
    this.smoothedBytesPerSecond,
  }) : transferredBytes = resumedBytes;

  factory TransferMetricsState.start({
    required int totalBytes,
    int resumedBytes = 0,
    DateTime? startedAt,
  }) {
    final now = startedAt ?? DateTime.now();
    return TransferMetricsState(
      totalBytes: totalBytes,
      resumedBytes: resumedBytes,
      startedAt: now,
      lastSampleAt: now,
      lastSampleBytes: resumedBytes,
    );
  }

  TransferMetricsUpdate update(int currentTransferredBytes, [DateTime? now]) {
    final nowTime = now ?? DateTime.now();
    transferredBytes = currentTransferredBytes.clamp(0, totalBytes);
    final elapsedMs = nowTime.difference(lastSampleAt).inMilliseconds;

    // Ignore samples shorter than ~250ms to avoid noisy values
    if (elapsedMs >= 250) {
      final bytesDelta = max(0, currentTransferredBytes - lastSampleBytes);
      final instantSpeed = (bytesDelta * 1000) / elapsedMs;

      if (smoothedBytesPerSecond == null) {
        smoothedBytesPerSecond = instantSpeed;
      } else {
        smoothedBytesPerSecond =
            (smoothedBytesPerSecond! * 0.7) + (instantSpeed * 0.3);
      }

      lastSampleAt = nowTime;
      lastSampleBytes = currentTransferredBytes;
    }

    final remainingBytes = max(0, totalBytes - currentTransferredBytes);
    final sessionTransferredBytes = max(0, currentTransferredBytes - resumedBytes);

    int? etaSeconds;
    if (smoothedBytesPerSecond != null && smoothedBytesPerSecond! > 0) {
      etaSeconds = (remainingBytes / smoothedBytesPerSecond!).round();
    }

    final progressFraction = totalBytes > 0
        ? (currentTransferredBytes / totalBytes).clamp(0.0, 1.0)
        : 0.0;

    return TransferMetricsUpdate(
      speedBps: smoothedBytesPerSecond,
      etaSeconds: etaSeconds,
      progressFraction: progressFraction,
      sessionTransferredBytes: sessionTransferredBytes,
      remainingBytes: remainingBytes,
    );
  }
}

class TransferMetricsUpdate {
  final double? speedBps;
  final int? etaSeconds;
  final double progressFraction;
  final int sessionTransferredBytes;
  final int remainingBytes;

  const TransferMetricsUpdate({
    required this.speedBps,
    required this.etaSeconds,
    required this.progressFraction,
    required this.sessionTransferredBytes,
    required this.remainingBytes,
  });
}

File: desktop/src/utils/transferMetrics.ts
export interface TransferMetricsState {
  resumedBytes: number;
  transferredBytes: number;
  totalBytes: number;
  startedAtMs: number;
  lastSampleAtMs: number;
  lastSampleBytes: number;
  smoothedBytesPerSecond: number | null;
}

export interface TransferMetricsUpdate {
  speedBps: number | null;
  etaSeconds: number | null;
  progressFraction: number;
  sessionTransferredBytes: number;
  remainingBytes: number;
}

export function createTransferMetrics(
  totalBytes: number,
  resumedBytes: number = 0,
  startedAtMs: number = Date.now()
): TransferMetricsState {
  return {
    resumedBytes,
    transferredBytes: resumedBytes,
    totalBytes,
    startedAtMs,
    lastSampleAtMs: startedAtMs,
    lastSampleBytes: resumedBytes,
    smoothedBytesPerSecond: null,
  };
}

export function updateTransferMetrics(
  state: TransferMetricsState,
  currentTransferredBytes: number,
  nowMs: number = Date.now()
): TransferMetricsUpdate {
  state.transferredBytes = Math.min(currentTransferredBytes, state.totalBytes);
  const timeDeltaSec = (nowMs - state.lastSampleAtMs) / 1000;

  // Ignore samples shorter than ~250ms to avoid noisy values
  if (timeDeltaSec >= 0.25) {
    const bytesDelta = Math.max(0, currentTransferredBytes - state.lastSampleBytes);
    const instantSpeed = bytesDelta / timeDeltaSec;

    if (state.smoothedBytesPerSecond === null) {
      state.smoothedBytesPerSecond = instantSpeed;
    } else {
      state.smoothedBytesPerSecond =
        state.smoothedBytesPerSecond * 0.7 + instantSpeed * 0.3;
    }

    state.lastSampleAtMs = nowMs;
    state.lastSampleBytes = currentTransferredBytes;
  }

  const remainingBytes = Math.max(0, state.totalBytes - currentTransferredBytes);
  const sessionTransferredBytes = Math.max(
    0,
    currentTransferredBytes - state.resumedBytes
  );

  let etaSeconds: number | null = null;
  if (state.smoothedBytesPerSecond && state.smoothedBytesPerSecond > 0) {
    etaSeconds = Math.round(remainingBytes / state.smoothedBytesPerSecond);
  }

  const progressFraction =
    state.totalBytes > 0
      ? Math.min(1.0, Math.max(0.0, currentTransferredBytes / state.totalBytes))
      : 0.0;

  return {
    speedBps: state.smoothedBytesPerSecond,
    etaSeconds,
    progressFraction,
    sessionTransferredBytes,
    remainingBytes,
  };
}
